godot-cell/src/guards.rs

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::ops::{Deref, DerefMut};
+use std::ptr::NonNull;
+use std::sync::Mutex;
+
+use crate::CellState;
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+
+/// Wraps a shared borrowed value of type `T`.
+///
+/// No other mutable borrows to the same value will be created while this guard exists.
+#[derive(Debug)]
+pub struct RefGuard<'a, T> {
+    state: &'a Mutex<CellState<T>>,
+    value: NonNull<T>,
+}
+
+impl<'a, T> RefGuard<'a, T> {
+    /// Create a new `GdRef` guard which can be immutably dereferenced.
+    ///
+    /// # Safety
+    ///
+    /// It must be safe to call [`as_ref()`](NonNull::as_ref) on `value` for as long as the guard is not
+    /// dropped.
+    ///
+    /// In particular you must ensure that:
+    ///
+    /// - The value behind the `value` pointer must be accessible for as long as the guard is not dropped.
+    /// - No new mutable references to the same value can be created.
+    /// - If there exist any other mutable references, then `value` must be derived from those references.
+    /// - Any existing mutable references must stop accessing this value while this guard exists.
+    pub(crate) unsafe fn new(state: &'a Mutex<CellState<T>>, value: NonNull<T>) -> Self {
+        Self { state, value }
+    }
+}
+
+impl<'a, T> Deref for RefGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: It is safe to call `as_ref()` on value because of the safety invariants of `new`.
+        unsafe { self.value.as_ref() }
```

Comment:
I'm not sure what you're saying here. If a is a consequence of b, then b is a cause of a. To me it seems the same to say "as a consequence of `new()`, it is safe to call `as_ref()`" and "because of `new()`, it is safe to call `as_ref()`"

Reply:
The wording of `# Safety` in `new()`:

> It must be safe to call [`as_ref()`](NonNull::as_ref) on `value` for as long as the guard is not dropped.

however sounds very much like "you must ensure that ... is safe to call" (i.e. precondition), and not a consequence (i.e. postcondition).  This could be reworded, and listed at the end (after the preconditions), along the lines "if you ensure above invariants, _then_ ...".